### PR TITLE
Support `editor.tabType`

### PR DESCRIPTION
**Pull request by benogle**
This does the simplest thing possible. When `editor.tabType` is set to `auto`, it does what it did before. When it is set to `hard` or `soft`, it uses those values all the time unless `setSoftTabs()` is explicitly called. The API is unchanged. Specs have been consolidated a bit as there were a couple places where the tab stuff was being tested.

Eventually (another PR), I'd like to change the name of `editor.softTabs` setting to `editor.defaultTabType` or something more clear. 

Closes https://github.com/atom/atom/pull/8186 @Abdillah, I've used a couple of your commits.
Closes https://github.com/atom/atom/issues/3719

### CODE REVIEW

File: src/text-editor.coffee [hunk 1]
```diff
@@ -176,9 +176,15 @@ class TextEditor extends Model
       @subscribe @displayBuffer.onDidAddDecoration (decoration) => @emit 'decoration-added', decoration
       @subscribe @displayBuffer.onDidRemoveDecoration (decoration) => @emit 'decoration-removed', decoration
 
+  subscribeToTabTypeConfig: ->
+    @tabTypeSubscription?.dispose()
```
Comment (benogle) [inline src/text-editor.coffee]: This is not super pretty. As I need to either add it to @disposables then remove on a grammar change, or explicitly destroy it in the `destroyed` method. I chose the latter

**Comment by maxbrunsfeld** [comment]
Do you think we should deprecate `editor.softTabs`?

**Comment by benogle** [comment]
One caveat with this is: if the user has explicitly set the soft tabs value to something and an event happens (config or grammar change), the user's setting is ignored and computed based on the rules of the setting. I thought it added extra complexity to support tracking the explicitly set choice, and I wasnt sure if it would be the correct behavior anyway.

**Comment by benogle** [comment]
> Do you think we should deprecate editor.softTabs?

We could change the name to `editor.defaultTabType` or something? Or yeah, we could deprecate it.

**Comment by benogle** [comment]
Anyone opposed to merging this?

**Comment by maxbrunsfeld** [comment]
:ship:

**Pull request merged**